VSDiagnostics/VSDiagnostics/VSDiagnostics.Test/Tests/General/ImplementEqualsAndGetHashCodeTests.cs

```diff
@@ -0,0 +1,241 @@
+namespace ConsoleApplication1
+{
+    class MyClass
+    {
+        string _foo = ""test"";
+        string _bar = ""test"";
+
+        string Foo { get; set; } = ""test"";
+        string Bar { get; set; } = ""test"";
+    }
+}";
+
+            var result = @"
+namespace ConsoleApplication1
+{
+    class MyClass
+    {
+        string _foo = ""test"";
+        string _bar = ""test"";
+
+        string Foo { get; set; } = ""test"";
+        string Bar { get; set; } = ""test"";
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null || GetType() != obj.GetType())
+            {
+                return false;
+            }
+
+            var value = (MyClass)obj;
+            return _foo == value._foo && _bar == value._bar && Foo == value.Foo && Bar == value.Bar;
+        }
+
+        public override int GetHashCode()
+        {
+            return _foo.GetHashCode() ^ _bar.GetHashCode() ^ Foo.GetHashCode() ^ Bar.GetHashCode();
+        }
+    }
+}";
+
+            VerifyDiagnostic(original, string.Format(ImplementEqualsAndGetHashCodeAnalyzer.Rule.MessageFormat.ToString(), "MyClass"));
+            VerifyFix(original, result);
+        }
+
+        [TestMethod]
+        public void ImplementEqualsAndGetHashCode_ClassDoesNotImplementEither_HasSetOnlyProperty()
+        {
+            var original = @"
+namespace ConsoleApplication1
+{
+    class MyClass
+    {
+        string _foo = ""test"";
+        string _bar = ""test"";
+
+        string Foo { get; set; } = ""test"";
+        string Bar
+        {
+            set { _bar = value; }
+        }
+    }
+}";
+
+            var result = @"
+namespace ConsoleApplication1
+{
+    class MyClass
+    {
+        string _foo = ""test"";
+        string _bar = ""test"";
+
+        string Foo { get; set; } = ""test"";
+        string Bar
+        {
+            set { _bar = value; }
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null || GetType() != obj.GetType())
+            {
+                return false;
+            }
+
+            var value = (MyClass)obj;
+            return _foo == value._foo && _bar == value._bar && Foo == value.Foo;
+        }
+
+        public override int GetHashCode()
+        {
+            return _foo.GetHashCode() ^ _bar.GetHashCode() ^ Foo.GetHashCode();
+        }
+    }
+}";
+
+            VerifyDiagnostic(original, string.Format(ImplementEqualsAndGetHashCodeAnalyzer.Rule.MessageFormat.ToString(), "MyClass"));
+            VerifyFix(original, result);
+        }
+
+        [TestMethod]
+        public void ImplementEqualsAndGetHashCode_ClassDoesNotImplementEither_DoesNotHaveFieldOrProperty()
+        {
+            var original = @"
+namespace ConsoleApplication1
+{
+    class MyClass
+    {
+    }
+}";
+
+            VerifyDiagnostic(original);
+        }
+
+        [TestMethod]
+        public void ImplementEqualsAndGetHashCode_ImplementsEquals_HasField()
```

Comment:
I think it would be good to still offer the code fix when one of the two methods has already been implemented. This will need a separate message of course